opm/simulators/wells/GasLiftRuntime_impl.hpp

```diff
@@ -0,0 +1,791 @@
+/*
+  Copyright 2020 Equinor ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <opm/simulators/wells/StandardWell.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/Schedule.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/Well/Well.hpp>
+#include <opm/simulators/utils/DeferredLogger.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/SummaryState.hpp>
+#include <opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/GasLiftOpt.hpp>
+
+#include <optional>
+#include <string>
+
+template<typename TypeTag>
+Opm::GasLiftRuntime<TypeTag>::
+GasLiftRuntime(
+    const StdWell &std_well,
+    const Simulator &ebos_simulator,
+    const SummaryState &summary_state,
+    DeferredLogger &deferred_logger,
+    std::vector<double> &potentials,
+    const WellState &well_state,
+    const Well::ProductionControls &controls
+) :
+    controls_{controls},
+    deferred_logger_{deferred_logger},
+    ebos_simulator_{ebos_simulator},
+    potentials_{potentials},
+    std_well_{std_well},
+    summary_state_{summary_state},
+    well_state_{well_state},
+    debug{false}  // extra debugging output
+{
+    int well_index = this->std_well_.indexOfWell();
+    const Well::ProducerCMode& control_mode
+        = well_state_.currentProductionControls()[well_index];
+    assert(control_mode == Well::ProducerCMode::THP );
+    const Opm::Schedule& schedule = this->ebos_simulator_.vanguard().schedule();
```

Comment:
The control_mode is also used for the asset. Try OPM_OPTIM_UNUSED

Reply:
> The `control_mode` is also used for the asset.

Did you mean `assert` ? I am not sure what you are referring to

Reply:
Yes, I mean assert. The control_mode is unused if not build in debug mode since the assertion will be skipped. 
```
const Well::ProducerCMode& OPM_OPTIM_UNUSED control_mode
        = well_state_.currentProductionControls()[well_index];
```
would make the compiler not warn against unused variable, I think. (not tested)

Reply:
Is there a good argument for `assert( ... ) ` rather than e.g.
```
if (control_mode != Well::ProducerCMode::THP)
     throw std::logic_error("Bug in flow - invalid control mode detected\n");
```

Reply:
Thanks for the comments! I have updated the PR.